query-algebrizer/src/types.rs

```diff
@@ -194,6 +194,88 @@ pub enum ColumnConstraint {
     HasType(TableAlias, ValueType),
 }
 
+#[derive(PartialEq, Eq, Debug)]
+pub enum ColumnConstraintOrAlternation {
+    Constraint(ColumnConstraint),
+    Alternation(ColumnAlternation),
+}
+
+impl From<ColumnConstraint> for ColumnConstraintOrAlternation {
+    fn from(thing: ColumnConstraint) -> Self {
+        ColumnConstraintOrAlternation::Constraint(thing)
+    }
+}
+
+/// An ordered collection of constraints that must all match.
+#[derive(PartialEq, Eq)]
+pub struct ColumnIntersection(pub Vec<ColumnConstraintOrAlternation>);
```

Comment:
Your data types aren't symmetric: `Intersection` is many `Constraint|Alteration`, but `Alteration` is many `Intersection`.  This can express many `Constraint`, but it's another layer of indirection.

Is this just an accident of how the code grew?  Is this intentional?  I.e., are you intending to normalize to a conjunction-of-disjunctions?  (Hope I got that right!)